File: code/packages/fsharp/json-serializer/JsonSerializer.fs
namespace CodingAdventures.JsonSerializer.FSharp

open System
open System.Globalization
open System.Text
open CodingAdventures.JsonValue.FSharp

type JsonSerializerError(message: string) =
    inherit Exception(message)

type SerializerConfig(?indentSize: int, ?indentChar: char, ?sortKeys: bool, ?trailingNewline: bool) =
    member _.IndentSize = defaultArg indentSize 2
    member _.IndentChar = defaultArg indentChar ' '
    member _.SortKeys = defaultArg sortKeys false
    member _.TrailingNewline = defaultArg trailingNewline false

[<AbstractClass; Sealed>]
type JsonSerializer private () =
    static member Serialize(value: JsonValue) =
        match value with
        | :? JsonNullValue -> "null"
        | :? JsonBooleanValue as booleanValue -> if booleanValue.Value then "true" else "false"
        | :? JsonNumberValue as numberValue -> JsonSerializer.SerializeNumber(numberValue)
        | :? JsonStringValue as stringValue -> JsonSerializer.SerializeString(stringValue.Value)
        | :? JsonArrayValue as arrayValue ->
            arrayValue.Elements |> Seq.map JsonSerializer.Serialize |> String.concat "," |> sprintf "[%s]"
        | :? JsonObjectValue as objectValue ->
            objectValue.Pairs
            |> Seq.map (fun pair -> sprintf "%s:%s" (JsonSerializer.SerializeString(pair.Key)) (JsonSerializer.Serialize(pair.Value)))
            |> String.concat ","
            |> sprintf "{%s}"
        | _ ->
            raise (JsonSerializerError(sprintf "Unsupported JsonValue node '%s'." (value.GetType().Name)))

    static member SerializePretty(value: JsonValue, ?config: SerializerConfig) =
        let resolved = defaultArg config (SerializerConfig())
        let text = JsonSerializer.SerializePrettyRecursive(value, resolved, 0)
        if resolved.TrailingNewline then text + "\n" else text

    static member Stringify(value: obj) =
        JsonValue.FromNative(value) |> JsonSerializer.Serialize

    static member StringifyPretty(value: obj, ?config: SerializerConfig) =
        JsonValue.FromNative(value) |> fun jsonValue -> JsonSerializer.SerializePretty(jsonValue, ?config = config)

    static member private SerializePrettyRecursive(value: JsonValue, config: SerializerConfig, depth: int) =
        let indentUnit = String.replicate config.IndentSize (string config.IndentChar)
        let currentIndent = String.Concat(Seq.replicate depth indentUnit)
        let nextIndent = String.Concat(Seq.replicate (depth + 1) indentUnit)

        match value with
        | :? JsonNullValue -> "null"
        | :? JsonBooleanValue as booleanValue -> if booleanValue.Value then "true" else "false"
        | :? JsonNumberValue as numberValue -> JsonSerializer.SerializeNumber(numberValue)
        | :? JsonStringValue as stringValue -> JsonSerializer.SerializeString(stringValue.Value)
        | :? JsonArrayValue as arrayValue ->
            if arrayValue.Elements.Count = 0 then
                "[]"
            else
                let lines =
                    arrayValue.Elements
                    |> Seq.map (fun element -> nextIndent + JsonSerializer.SerializePrettyRecursive(element, config, depth + 1))
                    |> String.concat ",\n"

                sprintf "[\n%s\n%s]" lines currentIndent
        | :? JsonObjectValue as objectValue ->
            if objectValue.Pairs.Count = 0 then
                "{}"
            else
                let pairs =
                    let basePairs = objectValue.Pairs |> Seq.toList
                    if config.SortKeys then
                        basePairs |> List.sortBy (fun pair -> pair.Key)
                    else
                        basePairs

                let lines =
                    pairs
                    |> Seq.map (fun pair ->
                        sprintf "%s%s: %s"
                            nextIndent
                            (JsonSerializer.SerializeString(pair.Key))
                            (JsonSerializer.SerializePrettyRecursive(pair.Value, config, depth + 1)))
                    |> String.concat ",\n"

                sprintf "{\n%s\n%s}" lines currentIndent
        | _ ->
            raise (JsonSerializerError(sprintf "Unsupported JsonValue node '%s'." (value.GetType().Name)))

    static member private SerializeNumber(numberValue: JsonNumberValue) =
        if Double.IsNaN(numberValue.Value) || Double.IsInfinity(numberValue.Value) then
            raise (JsonSerializerError("JSON does not support NaN or infinity."))

        if numberValue.IsInteger
           && numberValue.Value <= double Int64.MaxValue
           && numberValue.Value >= double Int64.MinValue
           && abs (numberValue.Value % 1.0) < Double.Epsilon then
            Convert.ToInt64(numberValue.Value, CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture)
        else
            numberValue.Value.ToString("R", CultureInfo.InvariantCulture)

    static member private SerializeString(value: string) =
        let builder = StringBuilder()
        builder.Append('"') |> ignore

        for ch in value do
            match ch with
            | '"' -> builder.Append("\\\"") |> ignore
            | '\\' -> builder.Append("\\\\") |> ignore
            | '\b' -> builder.Append("\\b") |> ignore
            | '\f' -> builder.Append("\\f") |> ignore
            | '\n' -> builder.Append("\\n") |> ignore
            | '\r' -> builder.Append("\\r") |> ignore
            | '\t' -> builder.Append("\\t") |> ignore
            | _ when Char.IsControl(ch) ->
                builder.Append("\\u") |> ignore
                builder.Append((int ch).ToString("x4", CultureInfo.InvariantCulture)) |> ignore
            | _ ->
                builder.Append(ch) |> ignore

        builder.Append('"') |> ignore
        builder.ToString()

File: code/packages/csharp/json-serializer/JsonSerializer.cs
namespace CodingAdventures.JsonSerializer;

using System.Globalization;
using System.Text;
using CodingAdventures.JsonValue;
using JsonNode = CodingAdventures.JsonValue.JsonValue;

public sealed class JsonSerializerError : Exception
{
    public JsonSerializerError(string message) : base(message)
    {
    }
}

public sealed record SerializerConfig(
    int IndentSize = 2,
    char IndentChar = ' ',
    bool SortKeys = false,
    bool TrailingNewline = false);

public static class JsonSerializer
{
    public static string Serialize(JsonNode value)
    {
        return value switch
        {
            JsonNullValue => "null",
            JsonBooleanValue booleanValue => booleanValue.Value ? "true" : "false",
            JsonNumberValue numberValue => SerializeNumber(numberValue),
            JsonStringValue stringValue => SerializeString(stringValue.Value),
            JsonArrayValue arrayValue => "[" + string.Join(",", arrayValue.Elements.Select(Serialize)) + "]",
            JsonObjectValue objectValue => "{" + string.Join(",", objectValue.Pairs.Select(pair => $"{SerializeString(pair.Key)}:{Serialize(pair.Value)}")) + "}",
            _ => throw new JsonSerializerError($"Unsupported JsonValue node '{value.GetType().Name}'."),
        };
    }

    public static string SerializePretty(JsonNode value, SerializerConfig? config = null)
    {
        var resolved = config ?? new SerializerConfig();
        var text = SerializePrettyRecursive(value, resolved, depth: 0);
        return resolved.TrailingNewline ? $"{text}\n" : text;
    }

    public static string Stringify(object? value) => Serialize(JsonNode.FromNative(value));

    public static string StringifyPretty(object? value, SerializerConfig? config = null)
    {
        return SerializePretty(JsonNode.FromNative(value), config);
    }

    private static string SerializePrettyRecursive(JsonNode value, SerializerConfig config, int depth)
    {
        var indentUnit = new string(config.IndentChar, config.IndentSize);
        var currentIndent = Repeat(indentUnit, depth);
        var nextIndent = Repeat(indentUnit, depth + 1);

        return value switch
        {
            JsonNullValue => "null",
            JsonBooleanValue booleanValue => booleanValue.Value ? "true" : "false",
            JsonNumberValue numberValue => SerializeNumber(numberValue),
            JsonStringValue stringValue => SerializeString(stringValue.Value),
            JsonArrayValue arrayValue => SerializePrettyArray(arrayValue, config, depth, currentIndent, nextIndent),
            JsonObjectValue objectValue => SerializePrettyObject(objectValue, config, depth, currentIndent, nextIndent),
            _ => throw new JsonSerializerError($"Unsupported JsonValue node '{value.GetType().Name}'."),
        };
    }

    private static string SerializePrettyArray(
        JsonArrayValue value,
        SerializerConfig config,
        int depth,
        string currentIndent,
        string nextIndent)
    {
        if (value.Elements.Count == 0)
        {
            return "[]";
        }

        var lines = value.Elements
            .Select(element => $"{nextIndent}{SerializePrettyRecursive(element, config, depth + 1)}");

        return $"[\n{string.Join(",\n", lines)}\n{currentIndent}]";
    }

    private static string SerializePrettyObject(
        JsonObjectValue value,
        SerializerConfig config,
        int depth,
        string currentIndent,
        string nextIndent)
    {
        if (value.Pairs.Count == 0)
        {
            return "{}";
        }

        var pairs = value.Pairs.AsEnumerable();
        if (config.SortKeys)
        {
            pairs = pairs.OrderBy(pair => pair.Key, StringComparer.Ordinal);
        }

        var lines = pairs.Select(pair =>
            $"{nextIndent}{SerializeString(pair.Key)}: {SerializePrettyRecursive(pair.Value, config, depth + 1)}");

        return $"{{\n{string.Join(",\n", lines)}\n{currentIndent}}}";
    }

    private static string SerializeNumber(JsonNumberValue number)
    {
        if (double.IsNaN(number.Value) || double.IsInfinity(number.Value))
        {
            throw new JsonSerializerError("JSON does not support NaN or infinity.");
        }

        if (number.IsInteger
            && number.Value <= long.MaxValue
            && number.Value >= long.MinValue
            && Math.Abs(number.Value % 1D) < double.Epsilon)
        {
            return Convert.ToInt64(number.Value, CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture);
        }

        return number.Value.ToString("R", CultureInfo.InvariantCulture);
    }

    private static string SerializeString(string value)
    {
        var builder = new StringBuilder();
        builder.Append('"');

        foreach (var ch in value)
        {
            switch (ch)
            {
                case '"':
                    builder.Append("\\\"");
                    break;
                case '\\':
                    builder.Append("\\\\");
                    break;
                case '\b':
                    builder.Append("\\b");
                    break;
                case '\f':
                    builder.Append("\\f");
                    break;
                case '\n':
                    builder.Append("\\n");
                    break;
                case '\r':
                    builder.Append("\\r");
                    break;
                case '\t':
                    builder.Append("\\t");
                    break;
                default:
                    if (char.IsControl(ch))
                    {
                        builder.Append("\\u");
                        builder.Append(((int)ch).ToString("x4", CultureInfo.InvariantCulture));
                    }
                    else
                    {
                        builder.Append(ch);
                    }

                    break;
            }
        }

        builder.Append('"');
        return builder.ToString();
    }

    private static string Repeat(string text, int count)
    {
        var builder = new StringBuilder();
        for (var index = 0; index < count; index++)
        {
            builder.Append(text);
        }

        return builder.ToString();
    }
}
